src/virtio/net.c

```diff
@@ -0,0 +1,385 @@
+        break;
+
+    case 1:
+        *ret_val = config->mac[4];
+        *ret_val |= config->mac[5] << 8;
+        break;
+
+    default:
+        LOG_NET_ERR("Unknown device config register: 0x%x\n", offset);
+        return 0;
+    }
+    return 1;
+}
+
+static int virtio_net_set_device_config(struct virtio_device *dev, uint32_t offset, uint32_t val)
+{
+    LOG_NET_ERR("All configuration fields are readonly\n");
+    return 0;
+}
+
+static void virtq_enqueue_used(struct virtq *virtq, uint32_t desc_head, uint32_t bytes_written)
+{
+    struct virtq_used_elem *used_elem = &virtq->used->ring[virtq->used->idx % virtq->num];
+    used_elem->id = desc_head;
+    used_elem->len = bytes_written;
+    virtq->used->idx++;
+}
+
+static void virtio_net_respond(struct virtio_device *dev)
+{
+    dev->data.InterruptStatus = BIT_LOW(0);
+    bool success = virq_inject(GUEST_VCPU_ID, dev->virq);
+    assert(success);
+}
+
+static void handle_tx_msg(struct virtio_device *dev,
+                          struct virtq *virtq,
+                          uint16_t desc_head,
+                          bool *notify_tx_server,
+                          bool *respond_to_client)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (net_queue_full_active(&state->tx)) {
+        LOG_NET_WRN("TX used ring full\n");
+        goto fail;
+    }
+
+    net_buff_desc_t sddf_buffer;
+    int error = net_dequeue_free(&state->tx, &sddf_buffer);
+    if (error) {
+        LOG_NET_WRN("Avail ring is empty\n");
+        goto fail;
+    }
+
+    void *dest_buf = state->tx_data + sddf_buffer.io_or_offset;
+
+    uint32_t written = 0;
+    uint32_t dest_remaining = NET_BUFFER_SIZE;
+
+    /* Strip virtio header before copying to sDDF */
+    uint32_t skip_remaining = sizeof(struct virtio_net_hdr_mrg_rxbuf);
+
+    struct virtq_desc *desc = &virtq->desc[desc_head];
+
+    while (dest_remaining > 0) {
+        uint32_t skipping = 0;
+        /* Work out how much of this descriptor must be skipped */
+        skipping = MIN(skip_remaining, desc->len);
+        /* Truncate packets that are large than BUF_SIZE */
+        uint32_t writing = MIN(dest_remaining, desc->len - skipping);
+
+        memcpy(dest_buf + written, (void *)desc->addr + skipping, writing);
+
+        skip_remaining -= skipping;
+        written += writing;
+        dest_remaining -= writing;
+
+        if (desc->flags & VIRTQ_DESC_F_NEXT) {
+            desc = &virtq->desc[desc->next];
+        } else {
+            break;
+        }
+    }
+
+    sddf_buffer.len = written;
+    error = net_enqueue_active(&state->tx, sddf_buffer);
+    /* This cannot fail as we check above */
+    assert(!error);
+
+    virtq_enqueue_used(virtq, desc_head, written);
+    *respond_to_client = true;
+    *notify_tx_server = true;
+    return;
+
+fail:
+    virtq_enqueue_used(virtq, desc_head, 0);
+    *respond_to_client = true;
+}
+
+static int virtio_net_queue_notify(struct virtio_device *dev)
```

Comment:
This should also check whether the device is ready. I think it's really the MMIO layer `src/virtio/mmio.c` that should be doing this before calling any of these callbacks, but it would be could to have an assert here as well.

Reply:
I've added the check to `net.c` in https://github.com/au-ts/libvmm/pull/95/commits/03091e3c2d99b3c0554d718033ce88fbbae4672b. We can change this to an assert later when its implemented in `mmio.c`